### PR TITLE
`PrescribedController` bug fixes

### LINKED ISSUE #3749: Maintain support for workflow using `PrescribedController::prescribeControlForActuator()` with model and controller loaded from file
A workflow for `PrescribedController` is to load a model with a controller and update a `Function` using `prescribeControlForActuator()`. Updating the `PrescribedContoller` class to use sockets has changed how this use case works, as `prescribeControlForActuator()` appends the function to the end of the `ControlFunctions` property, which can cause a couple issues (files attached that can be used to replicate these):
1. Appending an additional function means that the number of control functions is not equal to the number of socket connectees, throwing [here](https://github.com/opensim-org/opensim-core/blob/main/OpenSim/Simulation/Control/PrescribedController.cpp#L169).
2. In the case where the `Function` name is already the actuator name, using `prescribeControlForActuator()` with the actuator name will still append another `Function` (and now `ControlFunctions` will have 2 `Functions`s of the same name)

[PrescribedControllerExample.zip](https://github.com/opensim-org/opensim-core/files/14728746/PrescribedControllerExample.zip) using `4.5-2024-03-11-765d625`

A couple of ideas chatted about:
1. Relax the check (to allow for more control functions than actuators.
2. `prescribeControlForActuator()` could try to use the map to see if an actuator exists and update the associated `Function` (I'm a little surprised this doesn't happen [here](https://github.com/opensim-org/opensim-core/blob/main/OpenSim/Simulation/Control/PrescribedController.cpp#L198), perhaps there's a mix between just using the actuator name vs full path in the map?)

### PR DESCRIPTION
**Pull request by nickbianco**
Fixes issue #3749, #3750

### Brief summary of changes

These changes fix bugs in `PrescribedController` recently introduced after converting `Controller` to support a list `Socket` to maintain actuator connections. This PR also changes the behavior of `prescribeControlForActuator()` by making a copy of the `Function` at the passed in `Function` pointer rather than taking ownership. This change protects users from possible crashes when `Function` ownership is passed to the controller (e.g., during garbage collection in Java/Matlab).

The main changes:
- ~~A check was added during deserialization to ensure that the number of actuators in the list `Socket` matches the number of control function.~~ EDIT: removed, this was causing serialization tests to fail.
- The map `_actuLabelsToControlFunctionIndexMap` is also updated during deserialization to account for any control functions existing the model.
- The functions in `ControlFunctions` are now reordered based on the order of actuators connected via the list `Socket` to provide the correct order for `computeControls()` and serialization. This actually fixes an undetected bug: the member variable `_actuIndexToControlFunctionIndexMap` was not being used within `computeControls()` as it should have been. However, reordering `ControlFunctions` makes `_actuIndexToControlFunctionIndexMap` unnecessary, so it has been removed.

### Testing I've completed
- Added relavant tests to `testControllers.cpp`.
- Tested the changes on the sample scripts included in the related issues: no crashes occur when passing functions to `prescribeControlForActuator`.

### Looking for feedback on...
- If we're going to copy the `Function` passed via `prescribeControlForActuator()`, should change the signature to use a const reference to the function (i.e., `const Function& prescribedFunction`), rather than a pointer? It would change usage in C++, but scripting users wouldn't notice a difference.

### CHANGELOG.md (choose one)

- no need to update because...bug fixes (unless we change the signature for `prescribeControlForActuator()`).

- - -
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3781)

### CODE REVIEW

**Comment by nickbianco** [comment]
@carmichaelong this is now ready for review.

**Review by carmichaelong — CHANGES_REQUESTED** [review]
looking good, especially with the test coverage getting better. just some minor comments

> Reviewed 25 of 25 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3781#-:-NxTE3Jq7_5aVEyHrODM:b-y4p0as)* status: all files reviewed, 4 unresolved discussions (waiting on @nickbianco)

___
*[`OpenSim/Actuators/Test/testMuscles.cpp` line 291 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3781#-NxTCH6CAe2J-8solBhI:-NxTCH6CAe2J-8solBhJ:bc5mo9t) ([raw file](https://github.com/opensim-org/opensim-core/blob/891c72322b3cf29ac087fef48b90ac29c66489a7/OpenSim/Actuators/Test/testMuscles.cpp#L291)):*
> ```cpp
>         //for the prescribed muscle controller
>         muscleController->prescribeControlForActuator("muscle", 
>                 *control->clone());
> ```

perhaps remove `clone()` here?
___
*[`OpenSim/Analyses/Test/testOutputReporter.cpp` line 140 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3781#-NxTC5m2-tx5MD1LSkLi:-NxTC5m2-tx5MD1LSkLj:bc5mo9t) ([raw file](https://github.com/opensim-org/opensim-core/blob/891c72322b3cf29ac087fef48b90ac29c66489a7/OpenSim/Analyses/Test/testOutputReporter.cpp#L140)):*
> ```cpp
>     // Set the individual muscle control functions 
>     //for the prescribed muscle controller
>     muscleController->prescribeControlForActuator("muscle", *control.clone());
> ```

perhaps remove `clone()` here?
___
*[`OpenSim/Simulation/Control/PrescribedController.h` line 39 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3781#-NxTDOut4ZYUMWE7TVFF:-NxP1_7Z8jAZvtKCJq9K:b-78ha5q) ([raw file](https://github.com/opensim-org/opensim-core/blob/891c72322b3cf29ac087fef48b90ac29c66489a7/OpenSim/Simulation/Control/PrescribedController.h#L39)):*
> ```objc
>  * prescribe the control values of its actuators as a function of time.
>  *
>  * The control functions are specified in the `ControlFunctions` property. The
> ```

as chatted in person:
- in some cases the order matters (e.g., deserializing from xml) and others not (e.g., using `prescribeControlForAcutator()`
- i'm not sure how to best indicate this, so i'm ok with the current wording since if a user thinks they should be careful with ordering even when using `prescribeControlForAcutator()`, then it's still ok

tl;dr if you can come up with a succinct way to explain this, great. if not, i wouldn't consider this blocking
___
*[`OpenSim/Simulation/Control/PrescribedController.cpp` line 116 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3781#-NxP3y7bEDfn3pnDKWHr:-NxP3y7bEDfn3pnDKWHs:b3fdwzy) ([raw file](https://github.com/opensim-org/opensim-core/blob/891c72322b3cf29ac087fef48b90ac29c66489a7/OpenSim/Simulation/Control/PrescribedController.cpp#L116)):*
> ```cpp
>         controlFuncs.getSize(), socket.getNumConnectees());
> 
>     // If the 'ALL' keyword was provide via the actuator list (pre-4.6), then 
> ```

provide -> provided

**Review by nickbianco — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3781)* status: 21 of 25 files reviewed, 4 unresolved discussions (waiting on @carmichaelong)

___
*[`OpenSim/Actuators/Test/testMuscles.cpp` line 291 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3781#-NxTCH6CAe2J-8solBhI:-NxTXHlN-8bzjUR_QVd3:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/891c72322b3cf29ac087fef48b90ac29c66489a7/OpenSim/Actuators/Test/testMuscles.cpp#L291)):*
<details><summary><i>Previously, carmichaelong wrote…</i></summary><blockquote>

perhaps remove `clone()` here?
</blockquote></details>

Done.
___
*[`OpenSim/Analyses/Test/testOutputReporter.cpp` line 140 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3781#-NxTC5m2-tx5MD1LSkLi:-NxTXOzgCfZdjbsi5_yx:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/891c72322b3cf29ac087fef48b90ac29c66489a7/OpenSim/Analyses/Test/testOutputReporter.cpp#L140)):*
<details><summary><i>Previously, carmichaelong wrote…</i></summary><blockquote>

perhaps remove `clone()` here?
</blockquote></details>

Done.
___
*[`OpenSim/Simulation/Control/PrescribedController.h` line 39 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3781#-NxTDOut4ZYUMWE7TVFF:-NxTZUQdBEv27V_g3vz1:b-iyfzbv) ([raw file](https://github.com/opensim-org/opensim-core/blob/891c72322b3cf29ac087fef48b90ac29c66489a7/OpenSim/Simulation/Control/PrescribedController.h#L39)):*
<details><summary><i>Previously, carmichaelong wrote…</i></summary><blockquote>

as chatted in person:
- in some cases the order matters (e.g., deserializing from xml) and others not (e.g., using `prescribeControlForAcutator()`
- i'm not sure how to best indicate this, so i'm ok with the current wording since if a user thinks they should be careful with ordering even when using `prescribeControlForAcutator()`, then it's still ok

tl;dr if you can come up with a succinct way to explain this, great. if not, i wouldn't consider this blocking
</blockquote></details>

I took a stab at clarifying when order matters and when it doesn't.
___
*[`OpenSim/Simulation/Control/PrescribedController.cpp` line 116 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3781#-NxP3y7bEDfn3pnDKWHr:-NxTZTKY-AS5GJoRWoRx:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/891c72322b3cf29ac087fef48b90ac29c66489a7/OpenSim/Simulation/Control/PrescribedController.cpp#L116)):*
<details><summary><i>Previously, carmichaelong wrote…</i></summary><blockquote>

provide -> provided
</blockquote></details>

Done.

**Comment by nickbianco** [comment]
@carmichaelong ready for re-review.

**Review by carmichaelong — APPROVED** [review]
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3781)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @nickbianco)

**Comment by nickbianco** [comment]
Thank you @carmichaelong!
